caas/ecs/application.go

```diff
@@ -0,0 +1,640 @@
+	"github.com/juju/juju/environs/tags"
+	jujustorage "github.com/juju/juju/storage"
+)
+
+var (
+	jujuDataDir = paths.DataDir(paths.OSUnixLike)
+)
+
+type app struct {
+	name           string
+	clusterName    string
+	controllerUUID string
+	modelUUID      string
+	modelName      string
+	deploymentType caas.DeploymentType
+	client         ecsiface.ECSAPI
+	clock          clock.Clock
+}
+
+func newApplication(
+	name string,
+	clusterName string,
+	controllerUUID string,
+	modelUUID string,
+	modelName string,
+	deploymentType caas.DeploymentType,
+	client ecsiface.ECSAPI,
+	clock clock.Clock,
+) caas.Application {
+	return &app{
+		name:           name,
+		clusterName:    clusterName,
+		controllerUUID: controllerUUID,
+		modelUUID:      modelUUID,
+		modelName:      modelName,
+		deploymentType: deploymentType,
+		client:         client,
+		clock:          clock,
+	}
+}
+
+func (a *app) labels(extra map[string]string) map[string]*string {
+	tags := tags.ResourceTags(
+		names.NewModelTag(a.modelUUID),
+		names.NewControllerTag(a.controllerUUID),
+		// TODO(ecs): support model config ResourceTags().
+	)
+	for k, v := range extra {
+		tags[k] = v
+	}
+	return aws.StringMap(tags)
+}
+
+func (a *app) resourceName() string {
+	return fmt.Sprintf("%s-%s", a.modelName, a.name)
+}
+
+// Delete deletes the specified application.
+func (a *app) Delete() error {
+	// TODO(ecs)
+	return nil
+}
+
+func strPtrSlice(in ...string) (out []*string) {
+	for _, v := range in {
+		out = append(out, aws.String(v))
+	}
+	return out
+}
+
+func (a *app) volumeName(storageName string) string {
+	// ecs.Volume is not really a resource needs to be created, so we don't need use model name (resourceName()) to prefix it.
+	return fmt.Sprintf("%s-%s", a.name, storageName)
+}
+
+// getMountPathForFilesystem returns mount path.
+func getMountPathForFilesystem(idx int, appName string, fs jujustorage.KubernetesFilesystemParams) string {
+	if fs.Attachment != nil {
+		return fs.Attachment.Path
+	}
+	return fmt.Sprintf(
+		"%s/fs/%s/%s/%d",
+		jujuDataDir,
+		appName, fs.StorageName, idx,
+	)
+}
+
+func (a *app) handleFileSystems(filesystems []jujustorage.KubernetesFilesystemParams) (vols []*ecs.Volume, mounts []*ecs.MountPoint, err error) {
+	vols = make([]*ecs.Volume, len(filesystems))
+	mounts = make([]*ecs.MountPoint, len(filesystems))
+
+	volNames := set.NewStrings()
+	for idx, fs := range filesystems {
+		if volNames.Contains(fs.StorageName) {
+			return nil, nil, errors.NotValidf("duplicated volume %q", fs.StorageName)
+		}
+		volNames.Add(fs.StorageName)
+
+		ebsCfg, err := newEbsConfig(fs.Attributes)
+		if err != nil {
+			// This should never happen because it's been validated `storageProvider.ValidateConfig`.
+			return nil, nil, errors.NotValidf("storage attribute for %q", fs.StorageName)
+		}
+		vol := &ecs.Volume{
+			Name: aws.String(a.volumeName(fs.StorageName)),
+			DockerVolumeConfiguration: &ecs.DockerVolumeConfiguration{
+				Scope:         aws.String("shared"),
```

Comment:
These volumes are configured to `shared` for all the tasks for now.
We will need to test how to set here to `task` scope, so each task has its own volume.

Reply:
According to the docs: ```The scope for the Docker volume, which determines its lifecycle. Docker volumes that are scoped to a task are automatically provisioned when the task starts destroyed when the task is cleaned up. Docker volumes that are scoped as shared persist after the task stops.```. 

Does this mean that the storage gets reclaimed when a task gets moved to another node or clean up only applies when the task is _permanently_ destroyed (e.g. we scale down)?

Reply:
The ECS tasks are never restarted but always recreated if any `essential` containers died.
So it seems there is no point to keep the volume after the task died because there will be no chance to be re-attached again later at some point.
So we will need to test what happens it sets to the `task` scope.